script/tool/lib/src/custom_test_command.dart

```diff
@@ -0,0 +1,77 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file/file.dart';
+import 'package:platform/platform.dart';
+
+import 'common/package_looping_command.dart';
+import 'common/process_runner.dart';
+import 'common/repository_package.dart';
+
+const String _scriptName = 'run_tests.dart';
+const String _legacyScriptName = 'run_tests.sh';
+
+/// A command to run custom, package-local tests on packages.
+///
+/// This is an escape hatch for adding tests that this tooling doesn't support.
+/// It should be used sparingly; prefer instead to add functionality to this
+/// tooling to eliminate the need for bespoke tests.
+class CustomTestCommand extends PackageLoopingCommand {
+  /// Creates a custom test command instance.
+  CustomTestCommand(
+    Directory packagesDir, {
+    ProcessRunner processRunner = const ProcessRunner(),
+    Platform platform = const LocalPlatform(),
+  }) : super(packagesDir, processRunner: processRunner, platform: platform);
+
+  @override
+  final String name = 'custom-test';
+
+  @override
+  final String description = 'Runs package-specific custom tests defined in '
+      'a package\'s tool/$_scriptName file.\n\n'
+      'This command requires "dart" to be in your path.';
+
+  @override
+  Future<PackageResult> runForPackage(RepositoryPackage package) async {
+    final File script =
+        package.directory.childDirectory('tool').childFile(_scriptName);
```

Comment:
Pigeon has this in `bin`; I'll move it when I switch flutter/packages to this new command. Dart package layout docs say that `bin` is for package-client-facing commands, and scripts intended for package developers (the call out test scripts as an explicit example) belong in `tool`.